src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+      circpad_machine_remove_higher_token(mi, target_bin_us);
+      break;
+    case CIRCPAD_TOKEN_REMOVAL_EXACT:
+      circpad_machine_remove_exact(mi, target_bin_us);
+      break;
+  }
+
+  /* Check our token and state length limits */
+  return circpad_check_token_supply(mi);
+}
+
+/**
+ * Send a relay command with a relay cell payload on a circuit to
+ * the particular hopnum.
+ *
+ * Hopnum starts at 1 (1=guard, 2=middle, 3=exit, etc).
+ *
+ * Payload may be null.
+ *
+ * Returns negative on error, 0 on success.
+ */
+static int
+circpad_send_command_to_hop(origin_circuit_t *circ, int hopnum,
+                            uint8_t relay_command, const uint8_t *payload,
+                            ssize_t payload_len)
+{
+  crypt_path_t *target_hop = circuit_get_cpath_hop(circ, hopnum);
+  int ret;
+
+  /* Check that the cpath has the target hop */
+  if (!target_hop) {
+    log_fn(LOG_WARN,LD_CIRC,
+           "Padding circuit %u has %d hops, not %d",
+           circ->global_identifier,
+           circuit_get_cpath_len(circ), hopnum);
+    return -1;
+  }
+
+  /* Check that the target hop is opened */
+  if (target_hop->state != CPATH_STATE_OPEN) {
+    log_fn(LOG_WARN,LD_CIRC,
+           "Padding circuit %u has %d hops, not %d",
+           circ->global_identifier,
+           circuit_get_cpath_opened_len(circ), hopnum);
+    return -1;
+  }
+
+  /* Send the drop command to the second hop */
+  ret = relay_send_command_from_edge(0, TO_CIRCUIT(circ), relay_command,
+                                     (const char*)payload, payload_len,
+                                     target_hop);
+  return ret;
+}
+
+/**
+ * Callback helper to send a padding cell.
+ *
+ * This helper is called after our histogram-sampled delay period passes
+ * without another packet being sent first. If a packet is sent before this
+ * callback happens, it is canceled. So when we're called here, send padding
+ * right away.
+ */
+circpad_decision_t
+circpad_send_padding_cell_for_callback(circpad_machineinfo_t *mi)
+{
+  circuit_t *circ = mi->on_circ;
+  int machine_idx = mi->machine_index;
+  mi->padding_scheduled_at_us = 0;
+  circpad_statenum_t state = mi->current_state;
+
+  // Make sure circuit didn't close on us
+  if (mi->on_circ->marked_for_close) {
+    log_fn(LOG_INFO,LD_CIRC,
+           "Padding callback on a circuit marked for close. Ignoring.");
+    return CIRCPAD_STATE_CHANGED;
+  }
+
+  if (mi->histogram && mi->histogram_len) {
+    tor_assert(mi->chosen_bin < mi->histogram_len);
+    tor_assert(mi->histogram[mi->chosen_bin] > 0);
+    mi->histogram[mi->chosen_bin]--;
+  }
+
+  if (mi->state_length != CIRCPAD_STATE_LENGTH_INFINITE) {
+    tor_assert(mi->state_length > 0);
+    mi->state_length--;
+  }
+
+  /* Update padding counts */
+  mi->padding_sent++;
+  if (mi->padding_sent == UINT16_MAX) {
+    mi->padding_sent /= 2;
+    mi->nonpadding_sent /= 2;
+  }
+  circpad_global_padding_sent++;
+  if (circpad_global_padding_sent == UINT64_MAX) {
```

Comment:
If you sent a billion padding cells every second, it would take over 500 years to send this much padding.

Reply:
Verifying Nick's logic here:
```
five_years_in_secs = 5*365*86400
uint64_max = 0xffffffffffffffff
In [14]: uint64_max/five_years_in_secs
Out[14]: 116988483471L
```

Reply:
Clamping code removed in f413a343fbf58e1729d8bc3c35db0d2222e86c3e.